### PR TITLE
add editor pinning

### PR DESCRIPTION
**Pull request by kivistikliis**
---
### Purpose of this PR
Implemented editor pinning: updating and storing editor revisions in `_latest_editor_versions.metafile` on a nightly basis (3 types: staging `--fast`, latest_public `--published-only`, latest_internal)
- All jobs (referring to trunk) use the revisions from latest_internal. Editor priming for these is no longer required, because the revision build is kicked off already during the night
- All custom-revision jobs continue using editor priming (same way as before), to keep the resources free while waiting for the editor build

How editor pinning works is written under the yamato docs https://github.com/Unity-Technologies/Graphics/tree/yamato/editor-pin-clean/.yamato/docs#editor-priming-vs-editor-pinning 

---
### Testing status

**Manual Tests**: What did you do?
- Set it up also for 9.x.x to ensure that backporting this is safe

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)
Ran multiple jobs (both custom-revision and target revision on trunk):
- custom-revision ABV: https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252F_abv.yml%2523all_project_ci_CUSTOM-REVISION/3262935/job
- custom-revision (bits of nightly not covered in abv): android https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252Funiversal-android-opengles3.yml%2523Universal_Android_OpenGLES3_Standalone_CUSTOM-REVISION/3263012/job && https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252Funiversal-android-vulkan.yml%2523Universal_Android_Vulkan_Standalone_CUSTOM-REVISION/3263014/job
- custom-revision (bits of nightly not covered in abv): smoke tests https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252F_abv.yml%2523all_smoke_tests_CUSTOM-REVISION/3262981/job
- custom-revision (bits of nightly not covered in abv): trunk verification  https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252F_abv.yml%2523trunk_verification_CUSTOM-REVISION/3262992/job
- custom-revision pack and test all [package context]: https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin/.yamato%252F_packages.yml%2523all_package_ci_CUSTOM-REVISION/3264699/job
- custom-revision pack and test all [project context]: (IS PART OF ABV) https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin/.yamato%252F_projectcontext.yml%2523all_package_ci_project_CUSTOM-REVISION/3264704/job

- trunk nightly (covers ABV): https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252F_abv.yml%2523all_project_ci_nightly_trunk/3262844/job
- trunk pack and test all [package context]: https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252F_packages.yml%2523all_package_ci_trunk/3263110/job
- trunk pack and test all [project context]: (IS PART OF ABV)

- promote/publish dryrun  https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean/.yamato%252F_preview_publish.yml%2523promote_config_dry/3265309/job/pipeline
 

**Yamato**: (Select your branch):
- current 'clean' target branch https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-clean
- the 'in-progress' target branch https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin
- the 'in-progress' ci branch https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/yamato%252Feditor-pin-ci

---
### Comments to reviewers
Some of the jobs fail, but by investigating the logs can be seen that these are failing tests and getting the editor has succeeded

### CODE REVIEW

**Review by sophiaaar — APPROVED** [review]

File: .yamato/config/_latest_editor_versions.metafile
Comment (theo-pnv) [inline .yamato/config/_latest_editor_versions.metafile:4]: About the naming: Would it make sense to follow the "convention" of the Unity-QA-MultiTool software? 
- `trunk_latest_internal` would become `trunk_promoted`
- `trunk_latest_public` would become `trunk_published`
- I don't know if there's an equivalent for the staging one

File: .yamato/_abv.yml
Comment (theo-pnv) [inline .yamato/_abv.yml:17]: I am wondering if we can continue to retrieve the unity revision from an external file instead of having it hardcoded in the yaml. We would have to specify the file at this line (like before) and the `editor-pinning-merge-to-target` would commit this specific file only. 
This way there is no diff at this line when a dev wants to merge their branch with potential changes to this file into master (which potentially has a different editor version since the developer branched off of it). There is only a diff with the specific file.

**Review by theo-pnv — APPROVED** [review]
Looks good to me, great work!
The comments I left are no blocking the merge if we want to test this feature today. It's mainly questions and eventually ideas for future improvements on top of that.

File: .yamato/config/_latest_editor_versions.metafile
Comment (kivistikliis) [reply .yamato/config/_latest_editor_versions.metafile:4]: Good one! Right now it is kept consistent with the dots (the editor pinning is based on their solution), so this is a matter of preference. But is internal promoted and public published, as they seem to go opposite with packages convention?

File: .yamato/_abv.yml
Comment (kivistikliis) [reply .yamato/_abv.yml:17]: I like this idea! Definitely in for having less changes per these automated commits, but I am wondering how to do this since I'm not sure we can do it as before: before, editor-priming called the unity-downloader-cli to get the revision, and stored it in a file (i.e. this file is not in the repo, it gets created on the agent as Yamato runs). In current case (unless we create a job similar to priming), I think we would need to keep these files in the repo and have a file for each platform (i.e. `unity_revision_win.txt`, ...), where each of the files would essentially be a string extracted from the `_latest_editor_versions.metafile`. This could be done e.g. by a) another script which extracts latest_internal per each platform, and stores each of these in a separate file, or b) let editor pinning store versions straightaway in these `unity_revision_[platform].txt` files. Both of these options have their good and bad, but this could reduce to nr of changes being proportional to nr of platforms, instead of nr of ymls! Or perhaps there is another approach I didn't think of yet

Comment (kivistikliis) [reply .yamato/_abv.yml:17]: taking this thread to slack!

**Comment by sophiaaar** [comment]
osx standalone will be disabled in another PR - failures are unrelated to editor pinning

**Pull request merged**
